File: c++/Count the Number of Fair Pairs.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Solution {
public:
    long long countFairPairs(std::vector<int>& nums, int lower, int upper) {
        std::sort(nums.begin(), nums.end());
        
        int n = nums.size();
        long long count = 0;

        for (int i = 0; i < n; ++i) {
            long long leftBound = static_cast<long long>(lower) - nums[i];
            long long rightBound = static_cast<long long>(upper) - nums[i];

            int leftIndex = std::lower_bound(nums.begin() + i + 1, nums.end(), leftBound) - nums.begin();
            int rightIndex = std::upper_bound(nums.begin() + i + 1, nums.end(), rightBound) - nums.begin() - 1;

            if (leftIndex <= rightIndex) {
                count += (rightIndex - leftIndex + 1);
            }
        }

        return count;
    }
};
// Test cases
int main() {
    Solution solution;
    // case 1
    std::vector<int> nums1 = {0, 1, 7, 4, 4, 5};
    int lower1 = 3, upper1 = 6;
    std::cout << "Test Case 1 Output: " << solution.countFairPairs(nums1, lower1, upper1) << std::endl; // Expected: 6
    // case 2
    std::vector<int> nums2 = {1, 7, 9, 2, 5};
    int lower2 = 11, upper2 = 11;
    std::cout << "Test Case 2 Output: " << solution.countFairPairs(nums2, lower2, upper2) << std::endl; // Expected: 1

    return 0;
}

File: python/Count the Number of Fair Pairs.py
class Solution:
    """Count fair pairs with sum in [lower, upper]."""
    def countFairPairs(self, nums, lower, upper):
        nums.sort()
        count = 0
        n = len(nums)

        def lower_bound(arr, target, start):
            l, r = start, len(arr)
            while l < r:
                m = (l + r) // 2
                if arr[m] < target:
                    l = m + 1
                else:
                    r = m
            return l

        def upper_bound(arr, target, start):
            l, r = start, len(arr)
            while l < r:
                m = (l + r) // 2
                if arr[m] <= target:
                    l = m + 1
                else:
                    r = m
            return l

        for i in range(n):
            left = lower - nums[i]
            right = upper - nums[i]
            l = lower_bound(nums, left, i + 1)
            r = upper_bound(nums, right, i + 1)
            count += r - l
        return count
      
# Test cases
if __name__ == "__main__":
    solution = Solution()
  
    # case 1
    print(f"Test 1: {solution.countFairPairs([1,2,3,4,5], 3, 6)}")
  
    # case 2
    print(f"Test 2: {solution.countFairPairs([0,1,7,4,4,5], 3, 6)}")
  
    # case 3
    print(f"Test 3: {solution.countFairPairs([1,1,1,1], 2, 2)}")
